src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+    Duration::from_secs(3600),
+    Duration::from_secs(60),
+    Duration::from_secs(24 * 3600),
+);
+
+/// Limit on the amount of time a delegation is considered valid.
+const MAX_DELEGATION_VALIDITY: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(1_000_000),
+    Duration::from_secs(60),
+    Duration::from_secs(1_000_000_000),
+);
+
+// The following four flags are relevant to caching: AD, CD, DO, and RD.
+//
+// The AD flag needs to be part of the key when DO is clear. When replying,
+// if both AD and DO are not set in the original request then AD needs to be
+// cleared if it was set in the response (extra look up if no entry with
+// AD clear exists.
+//
+// The CD flag partitions the cache, responses to request with CD set must not
+// be visible to requests with CD and vice versa.
+//
+// A request with DO set can only be satisfied with a response to a request
+// with CD set. However, if CD in the request is clear then a reponse to a
+// request with CD set can be used if all unrequested DNSSEC records are
+// stripped.
+//
+// A request with RD clear can be satisfied by a response to a request with
+// RD set. For simplicitly requests with RD set can only get a response to
+// a request with RD set. In theory some responses to requests with RD clear
+// could satisfy requests with RD set.
+
+// Negative caching is described in RFC 2308.
+// NXDOMAIN and NODATA require special treatment. NXDOMAIN can be found
+// directly in the rcode field. NODATA is the condition where the answer
+// section does not contain any record that matches qtype and the message
+// is not a referral. NODATA is distinguished for a referral by the presence
+// of a SOA record in the authority section (a SOA record present implies
+// NODATA). A referral has one or more NS records in the authority section.
+// A NXDOMAIN reponse can only be cache if a SOA record is present in the
+// authority section. If the SOA record is absent then the NXDOMAIN response
+// should not be cached.
+// The TTL of the SOA record should reflect how long the response can be
+// cached. So no special treat is needed. Except that a different value should
+// limit the maximum time a negative response can be cached.
+//
+// Caching unreachable upstream should be limited to 5 minutes.
+// Caching SERVFAIL should be limited to 5 minutes.
+
+// RFC 8020 suggests a separate <QNAME, QCLASS> cache for NXDOMAIN, but
+// that may be too hard to implement.
+
+// RFC 9520 requires resolution failures to be cached for at least one
+// second. Resolution failure must not be cached for longer than 5 minutes.
+
+// RFC 8767 describes serving stale data.
+
+//------------ Config ---------------------------------------------------------
+
+/// Configuration of a cache.
+#[derive(Clone, Debug)]
+pub struct Config {
+    /// Maximum number of cache entries.
+    max_cache_entries: u64,
+
+    /// Maximum validity of a normal result.
```

Comment:
There are more comments on the constants, e.g. `MAX_VALIDITY`, than there are on these fields, but these fields comments appear in the public RustDoc while the comments on the constants do not. This is a bit strange and unhelpful.